From 1630a49b58937fd530ce2e9bfbba09aeeb647fe3 Mon Sep 17 00:00:00 2001
From: Alexander Duyck <alexander.h.duyck@redhat.com>
Date: Fri, 31 Oct 2014 19:06:45 -0400
Subject: [net] ip6_tunnel: Return an error when adding an existing tunnel

Message-id: <20141031190645.8251.79709.stgit@ahduyck-vm-rhel7>
Patchwork-id: 98658
O-Subject: [RHEL7.1 net PATCH 6/7] ip6_tunnel: Return an error when adding an existing tunnel.
Bugzilla: 1151886 1152368
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>
RH-Acked-by: Hannes Frederic Sowa <hannes@redhat.com>
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>

Bugzilla: 1152368
Brew Build: 8179440

Backport of upstream commit: 2b0bb01b6edb3e13c7f71e43bf3a173a795b7b66

 commit 2b0bb01b6edb3e13c7f71e43bf3a173a795b7b66
 Author: Steffen Klassert <steffen.klassert@secunet.com>
 Date:   Mon Sep 22 10:07:24 2014 +0200

    ip6_tunnel: Return an error when adding an existing tunnel.

    ip6_tnl_locate() should not return an existing tunnel if
    create is true. Otherwise it is possible to add the same
    tunnel multiple times without getting an error.

    So return NULL if the tunnel that should be created already
    exists.

    Signed-off-by: Steffen Klassert <steffen.klassert@secunet.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Alexander Duyck <alexander.h.duyck@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv6/ip6_tunnel.c b/net/ipv6/ip6_tunnel.c
index 4f42a0d..24825e9 100644
--- a/net/ipv6/ip6_tunnel.c
+++ b/net/ipv6/ip6_tunnel.c
@@ -355,8 +355,12 @@ static struct ip6_tnl *ip6_tnl_locate(struct net *net,
       (t = rtnl_dereference(*tp)) != NULL;
       tp = &t->next) {
   if (ipv6_addr_equal(local, &t->parms.laddr) &&
-      ipv6_addr_equal(remote, &t->parms.raddr))
+      ipv6_addr_equal(remote, &t->parms.raddr)) {
+   if (create)
+    return NULL;
+
    return t;
+  }
  }
  if (!create)
   return NULL;
-- 
1.7.1